etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,187 @@
+        self._level_to_index: Dict[str, int] = {level_names[i]: i for i in range(tree_depth)}
+
+        self._sub_segment_size_map: Dict[str, int] = {k: len(v) for k, v in level_structure.items()}
+
+        self._segment_subtree_size_map: Dict[str, int] = self._get_subtree_sizes(hierarchy_levels)
+
+        self._segment_to_level: Dict[str, str] = {
+            segment: level for level in self._level_series for segment in self._level_series[level]
+        }
+
+    def _find_graph_structure(self, adj_list: Dict[str, List[str]]):
+        """Find hierarchy top level (root of tree)."""
+        children = set(chain(*adj_list.values()))
+        parents = set(adj_list.keys())
+
+        tree_roots = parents.difference(children)
+        if len(tree_roots) != 1:
+            raise ValueError("Invalid tree definition: unable to find root!")
+
+        self._hierarchy_interm_nodes = parents & children
+        self._hierarchy_leaves = children.difference(parents)
+
+        tree_root = tree_roots.pop()
+        self._hierarchy_root = tree_root
+
+    def _find_hierarchy_levels(self, hierarchy_structure: Dict[str, List[str]]):
+        """Traverse hierarchy tree to group segments into levels."""
+        nodes: Set[str] = self._hierarchy_interm_nodes | self._hierarchy_leaves
+        nodes.add(str(self._hierarchy_root))
+
+        num_edges = sum(map(len, hierarchy_structure.values()))
+
+        num_nodes = len(nodes)
+        if num_edges != num_nodes - 1:
+            raise ValueError("Invalid tree definition: invalid number of nodes and edges!")
+
+        leaves_level = None
+        node_levels = []
+        seen_nodes = {self._hierarchy_root}
+        queue: Queue = Queue()
+        queue.put((self._hierarchy_root, 0))
+        while not queue.empty():
+            node, level = queue.get()
+            node_levels.append((level, node))
+            child_nodes = hierarchy_structure.get(node, [])
+
+            if len(child_nodes) == 0:
+                if leaves_level is not None and level != leaves_level:
+                    raise ValueError("All hierarchy tree leaves must be on the same level!")
+                else:
+                    leaves_level = level
+
+            for adj_node in child_nodes:
+                if adj_node not in seen_nodes:
+                    queue.put((adj_node, level + 1))
+                    seen_nodes.add(adj_node)
+
+        if len(seen_nodes) != num_nodes:
+            raise ValueError("Invalid tree definition: disconnected graph!")
+
+        levels = defaultdict(list)
+        for level, node in node_levels:
+            levels[level].append(node)
+
+        return levels
+
+    def _get_subtree_sizes(self, hierarchy_levels: Dict[int, List[str]]) -> Dict[str, int]:
+        """Compute subtree size for each node."""
+        subtree_size_map = dict()
+        for _, node_list in sorted(hierarchy_levels.items(), key=lambda x: x[0], reverse=True):
```

Comment:
May be 
- `node_list` -> `level_nodes`
- `subtree_size_map` -> `subtree_size `